### PR TITLE
HV-455 The annotation processor should verify the groups parameter of a constraint declaration

### PR DESCRIPTION
**Pull request by marko-bekhta**
https://hibernate.atlassian.net/browse/HV-455

added check for groups parameter to verify that groups are represented by interfaces.

One more for annotation processor :)

### CODE REVIEW

File: annotation-processor/src/main/java/org/hibernate/validator/ap/checks/annotationparameters/AnnotationParametersGroupsCheck.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+		return true;
+	}
+
+	@Override
+	protected Set<ConstraintCheckError> doCheck(Element element, AnnotationMirror annotation) {
+		List<? extends AnnotationValue> annotationValue = annotationApiHelper.getAnnotationArrayValue( annotation, "groups" );
+		Set<ConstraintCheckError> issues = CollectionHelper.newHashSet();
+
+		for ( AnnotationValue value : annotationValue ) {
+			if ( value.getValue() instanceof DeclaredType && !( (DeclaredType) value.getValue() ).asElement().getKind().isInterface() ) {
```
Comment (gsmet) [inline annotation-processor/src/main/java/org/hibernate/validator/ap/checks/annotationparameters/AnnotationParametersGroupsCheck.java]: Wouldn't `!( value.getValue() instanceof DeclaredType ) || ...` be a better test?

File: annotation-processor/src/test/java/org/hibernate/validator/ap/testmodel/annotationparameters/ValidGroupsParameters.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+package org.hibernate.validator.ap.testmodel.annotationparameters;
+
+import javax.validation.constraints.Size;
+
+/**
+ * @author Marko Bekhta
+ */
+public class ValidGroupsParameters {
+
+
```
Comment (gsmet) [inline annotation-processor/src/test/java/org/hibernate/validator/ap/testmodel/annotationparameters/ValidGroupsParameters.java]: Unnecessary whitespace.

File: annotation-processor/src/main/java/org/hibernate/validator/ap/checks/annotationparameters/AnnotationParametersGroupsCheck.java (hunk 1 above)
Comment (marko-bekhta) [inline annotation-processor/src/main/java/org/hibernate/validator/ap/checks/annotationparameters/AnnotationParametersGroupsCheck.java]: hmm... this `instanceof` is more for safety of next conversion. I've added it more for the cases when the annotation is of some different kind and groups parameter is present but contains something else ... But if it's only for constraint annotations it can be probably be removed completely. WDYT ?

Comment (gunnarmorling) [inline annotation-processor/src/main/java/org/hibernate/validator/ap/checks/annotationparameters/AnnotationParametersGroupsCheck.java]: The current check implementation would miss if e.g. `int.class` was given as a group. I think the following would be right:

* Safely assume a `TypeMirror` is given as a value (`groups` only expects class objects, so the user couldn't provide a String or long or so)
* If `getKind()` is not DECLARED, flag it as error (preventing something like `int.class` to be given)
* If kind is DECLARED but the type is not an interface, flag it as error (as you already do)

**Comment by gunnarmorling** [comment]
Rebased and applied. Many thanks, @marko-bekhta!
